src/D2L.Bmx/OktaAuthenticator.cs

```diff
@@ -49,38 +48,35 @@ bool nonInteractive
 
 		_oktaApi.SetOrganization( org );
 
-		var cachedSession = await AuthenticateFromCacheAsync( org, user, _oktaApi );
-		if( cachedSession.SuccessfulAuthentication ) {
+		if( await TryAuthenticateFromCacheAsync( org, user, _oktaApi ) ) {
 			return _oktaApi;
-		} else if( nonInteractive ) {
+		}
+		if( nonInteractive ) {
 			throw new BmxException( "Authentication failed. No cached session" );
 		}
 
 		string password = _consolePrompter.PromptPassword();
 
-		var authState = await _oktaApi.AuthenticateAsync( new AuthenticateOptions( user, password ) );
-
-		string? sessionToken = null;
+		var authnResponse = await _oktaApi.AuthenticateAsync( user, password );
 
-		if( authState.OktaStateToken is not null ) {
-			OktaMfaFactor mfaFactor = _consolePrompter.SelectMfa( authState.OktaMfaFactors );
+		if( authnResponse is AuthenticateResponse.MfaRequired mfaInfo ) {
+			OktaMfaFactor mfaFactor = _consolePrompter.SelectMfa( mfaInfo.Factors );
 
 			if( !IsMfaFactorTypeSupported( mfaFactor.FactorType ) ) {
 				throw new BmxException( "Selected MFA not supported by BMX." );
 			}
 
 			// TODO: Handle retry
 			if( mfaFactor.FactorType is "sms" or "call" or "email" ) {
-				await _oktaApi.IssueMfaChallengeAsync( authState, mfaFactor );
+				await _oktaApi.IssueMfaChallengeAsync( mfaInfo.StateToken, mfaFactor.Id );
 			}
 			string mfaResponse = _consolePrompter.GetMfaResponse( mfaFactor.FactorType == "question" ? "Answer" : "PassCode" );
-			sessionToken = await _oktaApi.VerifyMfaChallengeResponseAsync( authState, mfaFactor, mfaResponse );
-		} else if( authState.OktaSessionToken is not null ) {
-			sessionToken = authState.OktaSessionToken;
+			authnResponse = await _oktaApi.VerifyMfaChallengeResponseAsync( mfaInfo.StateToken, mfaFactor.Id, mfaResponse );
```

Comment:
this is only possible because of the polymorphic return types, and enables simpler logic below at [L77-L78](https://github.com/Brightspace/bmx/pull/313/files#diff-640013b3685b8a2cbcee8c5b2e26b1afc631ea22dc293c44a6f402feab21ff73R77-R78)